kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/objcexport/ObjCExportNamer.kt

```diff
@@ -902,6 +929,84 @@ private fun ObjCExportMapper.canHaveSameName(first: PropertyDescriptor, second:
+        private val kotlinName: String,
+        private val objCName: String?,
+        private val swiftName: String?,
+        val isExact: Boolean
+) {
+    // TODO: Prevent mangling when objCName or swiftName is provided
+
+    fun asString(forSwift: Boolean): String = swiftName.takeIf { forSwift } ?: objCName ?: kotlinName
+
+    fun asIdentifier(forSwift: Boolean, default: () -> String = kotlinName::toIdentifier): String =
+            swiftName.takeIf { forSwift } ?: objCName ?: default()
+}
+
+private fun DeclarationDescriptor.getObjCName(): ObjCName {
+    var objCName: String? = null
+    var swiftName: String? = null
+    var isExact = false
+    annotations.findAnnotation(KonanFqNames.objCName)?.let { annotation ->
+        objCName = annotation.argumentValue("name")?.value?.cast()
+        swiftName = annotation.argumentValue("swiftName")?.value?.cast()
+        isExact = annotation.argumentValue("exact")?.value?.cast() ?: false
+    }
+    return ObjCName(name.asString(), objCName, swiftName, isExact)
+}
+
+private fun CallableDescriptor.getObjCName(): ObjCName =
+        overriddenDescriptors.firstOrNull()?.getObjCName() ?: (this as DeclarationDescriptor).getObjCName()
+
+private fun ParameterDescriptor.getObjCName(): ObjCName {
+    val callableDescriptor = containingDeclaration as? CallableDescriptor ?: return (this as CallableDescriptor).getObjCName()
+    fun CallableDescriptor.getBase(): CallableDescriptor = overriddenDescriptors.firstOrNull()?.getBase() ?: this
+    val baseCallableDescriptor = callableDescriptor.getBase()
+    if (callableDescriptor.extensionReceiverParameter == this) {
+        return (baseCallableDescriptor.extensionReceiverParameter as CallableDescriptor).getObjCName()
+    }
+    val parameterIndex = callableDescriptor.valueParameters.indexOf(this)
+    if (parameterIndex != -1) {
+        return (baseCallableDescriptor.valueParameters[parameterIndex] as CallableDescriptor).getObjCName()
+    }
+    error("Unexpected parameter: $this")
+}
+
+private val objCNameShortName = KonanFqNames.objCName.shortName().asString()
+
+private fun KtClassOrObject.getObjCName(): ObjCName {
+    var objCName: String? = null
+    var swiftName: String? = null
+    var isExact = false
+    annotationEntries.firstOrNull {
+        it.calleeExpression?.constructorReferenceExpression?.getReferencedName() == objCNameShortName
+    }?.let { annotation ->
+        fun ValueArgument.getStringValue(): String? {
+            val stringTemplateExpression = when (this) {
+                is KtValueArgument -> stringTemplateExpression
+                else -> getArgumentExpression() as? KtStringTemplateExpression
+            } ?: return null
+            return (stringTemplateExpression.entries.singleOrNull() as? KtLiteralStringTemplateEntry)?.text
+        }
+
+        fun ValueArgument.getBooleanValue(): Boolean =
+                (getArgumentExpression() as? KtConstantExpression)?.node?.text?.toBoolean() ?: false
```

Comment:
This code is used in IDE.
`toBoolean` will probably throw an exception if the string is not `true` or `false`.
This is not fine:
1. What if it is not a boolean literal, but a reference to a `const val`? Supporting this case would be complicated, but at least IDE shouldn't throw an exception in this case.
2. A user might have simply written incorrect code. For example, `@ObjCName("abc", exact = "def")`. IDE should highlight this as error, but not throw an exception.

Reply:
Yeah I wasn't sure if this was actually the correct way to get the boolean. 
We could use `toBooleanOrNull` to solve the second issue, but the first one seems harder to fix.

The following compiles fine, but ignores `name` and `exact`:
```kotlin
private const val name = "ObjCNameObjCObject"
private const val exact = true

@ObjCName(name, "ObjCNameSwiftObject", exact)
object ObjCNameKotlinObject
```

Do you have any pointers for this?

Reply:
> I hacked together the following: https://github.com/google/kotlin/commit/d0232bf080067eee0b77baf09e9a5a0add7f9274

Interesting. Where are the tests for the lazy exporter located?

Reply:
>  but the first one seems harder to fix.

Exactly, that's why I proposed not to support it in IDE for now :)

> I hacked together the following: https://github.com/google/kotlin/commit/d0232bf080067eee0b77baf09e9a5a0add7f9274
> @SvyatoslavScherbina does that seem like the right way to go? 

Hard to say, I'm not actually an expert in IDE. IIUC, this would mean that indexing would trigger resolution for every class, right? This might be a huge performance hit.

> I don't really see how the lazy objective-c exporter can get the right name unless we go through resolution for it?

I see a couple of options. For example, we could prohibit using anything but boolean literals for `exact` values, or make it a separate annotation (e.g. `@ExactObjCName` or `@ObjCName.Exact`).

> Interesting. Where are the tests for the lazy exporter located?

`expectedLazy.h` is the only test I'm aware of.

Reply:
Alright I think I get the issues now (correct me if I am wrong).

1. using `const val` references as values for the annotation isn't currently working and would require resolution.
I don't think there is a real use case for this (since the names and exact value are tightly coupled to a declaration). Meaning we could prohibit it, only allowing literal string and boolean values, correct?
2. exceptions are thrown in case the arguments aren't in the expected format.
In this case we should make sure to silently ignore such values.
The compiler will fail with an appropriate error message anyway.

> `expectedLazy.h` is the only test I'm aware of.

That one is passing, so we shouldn't have any issues if we prohibit the usage of reference values.

Reply:
> Alright I think I get the issues now (correct me if I am wrong).
> 
> 1. using `const val` references as values for the annotation isn't currently working and would require resolution.
>    I don't think there is a real use case for this (since the names and exact value are tightly coupled to a declaration). Meaning we could prohibit it, only allowing literal string and boolean values, correct?

Yes. I agree that it is probably fine to limit to string constants and boolean constants for now. If we would like to expand on it in the future, that should be relatively easy as well. In what I hacked up we do resolution always to get the name. However, we can see the `@ObjCName` in the PSI, so we could do resolution when getting the name only when that annotation is present. Then at least the cost would only be incurred for the declarations where it is necessary. I would be happy to do that if we end up seeing a need, but I would support just limiting to constants for now.

> 2. exceptions are thrown in case the arguments aren't in the expected format.
>    In this case we should make sure to silently ignore such values.
>    The compiler will fail with an appropriate error message anyway.

That is my understanding as well. If the code cannot be compiled, the editor does a best effort. :)

> > `expectedLazy.h` is the only test I'm aware of.
> 
> That one is passing, so we shouldn't have any issues if we prohibit the usage of reference values.

Reply:
`toBoolean` wouldn't actually throw an exception, but instead return `false`. Updated to `toBooleanStrictOrNull` to make the intended behaviour more clear.

The frontend checker will now prohibit non literal string and boolean values, making sure we don't need resolution in the lazy exporter.

I have also added a test case to verify that there aren't any exceptions in case the user writes invalid code.

Reply:
> Alright I think I get the issues now (correct me if I am wrong).

Totally correct, thank you!